src/java/main/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1054,6 +1054,8 @@ private void sendPing() {
         private boolean saslLoginFailed = false;
 
         private void startConnect() throws IOException {
+            // initializing it for new connection
+            saslLoginFailed = false;
```

Comment:
How about setting to false only after client has successfully logged in instead of setting to false at the beginning. 
```
zooKeeperSaslClient = new ZooKeeperSaslClient(getServerPrincipal(addr), clientConfig);
saslLoginFailed = false;
```
I'm not sure about any concurrency cases between narrow window of resetting the flag to false and the client attempt to do login operation. Since the previous operation was failed, the client should still reflect the status until the login is successful next time, right?